File: 673. Number of Longest Increasing Subsequence/Java/Solution.java
class Solution {
    public int findNumberOfLIS(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        int numsLength = nums.length;
        int[] lengths = new int[numsLength];
        int[] counts = new int[numsLength];
        int maxLength = Integer.MIN_VALUE;
        for (int i = 0; i < numsLength; i++) {
            lengths[i] = 1;
            counts[i] = 1;
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    if (lengths[i] == lengths[j] + 1) {
                        counts[i] += counts[j];
                    } else if (lengths[i] < lengths[j] + 1) {
                        lengths[i] = lengths[j] + 1;
                        counts[i] = counts[j];
                    }
                }
            }
            maxLength = Math.max(maxLength, lengths[i]);
        }
        int result = 0;
        for (int i = 0; i < numsLength; i++) {
            if (lengths[i] == maxLength) {
                result += counts[i];
            }
        }
        return result;
    }
}

File: 673. Number of Longest Increasing Subsequence/Python/Solution.py
class Solution:
    def findNumberOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0
        nums_length = len(nums)
        lengths = [1] * nums_length
        counts = [1] * nums_length
        for i in range(nums_length):
            for j in range(i):
                if nums[i] > nums[j]:
                    if lengths[i] == lengths[j] + 1:
                        counts[i] += counts[j]
                    elif lengths[i] < lengths[j] + 1:
                        lengths[i] = lengths[j] + 1
                        counts[i] = counts[j]
        max_length = max(lengths)
        result = 0
        for i in range(nums_length):
            if lengths[i] == max_length:
                result += counts[i]
        return result
